fendermint/vm/topdown/src/finality/null.rs

```diff
@@ -137,18 +120,80 @@ impl FinalityWithNull {
 }
 
 impl FinalityWithNull {
+    /// Returns the number of blocks cached.
+    pub(crate) fn cached_blocks(&self) -> Stm<BlockHeight> {
+        let cache = self.cached_data.read()?;
+        Ok(cache.size() as BlockHeight)
+    }
+
     pub(crate) fn block_hash_at_height(&self, height: BlockHeight) -> Stm<Option<BlockHash>> {
+        if let Some(f) = self.last_committed_finality.read()?.as_ref() {
+            if f.height == height {
+                return Ok(Some(f.block_hash.clone()));
+            }
+        }
+
         self.get_at_height(height, |i| i.0.clone())
     }
 
-    pub(crate) fn latest_height(&self) -> Stm<Option<BlockHeight>> {
+    pub(crate) fn latest_height_in_cache(&self) -> Stm<Option<BlockHeight>> {
         let cache = self.cached_data.read()?;
         Ok(cache.upper_bound())
     }
+
+    /// Get the latest height tracked in the provider, includes both cache and last committed finality
+    pub(crate) fn latest_height(&self) -> Stm<Option<BlockHeight>> {
+        let h = if let Some(h) = self.latest_height_in_cache()? {
+            h
+        } else if let Some(p) = self.last_committed_finality()? {
+            p.height
+        } else {
+            return Ok(None);
+        };
+        Ok(Some(h))
+    }
 }
 
 /// All the private functions
 impl FinalityWithNull {
+    /// Get the first non-null block in the range [start, end].
+    fn min_nonnull_block(&self, start: BlockHeight, end: BlockHeight) -> Stm<Option<BlockHeight>> {
+        let cache = self.cached_data.read()?;
+        for h in start..=end {
+            if let Some(Some(_)) = cache.get_value(h) {
+                return Ok(Some(h));
+            }
+        }
+        Ok(None)
+    }
+
+    fn propose_next_height(&self) -> Stm<Option<BlockHeight>> {
```

Comment:
We shouldn't use an interval and always propose a specific height because this gives us nothing. Ideally we should: 
- Propose whatever I see as the latest height as long as it is not higher than `prev_committed_finality + MAX_DELTA`. If this height is null, we always search back. 
 
`MAX_DELTA` for now should big enough so it ensure at least one non-null block in that window (we can address small MAX_DELTA with all null-rounds in the future). For now, if everything null let's crash it because it breaks our assumption it would mean there is something really wrong in the parent.

Reply:
Merge this with Akosh's proposal of having an additional delay from proposal to acceptance: https://github.com/consensus-shipyard/fendermint/pull/435/files